### ISSUE TITLE
[new rule] prefer jest.doMock

### ISSUE THREAD

**Issue by chrisblossom** [jest-community/eslint-plugin-jest#283]
I have found using `jest.doMock` over `jest.mock` works better because it is predictable and works well with `resetMocks`, `resetModules` and `restoreMocks`.

**Comment by SimenB** [comment]
It won't work with `import` as they're hoisted to the top level. I'm not against adding a rule for it, though 🙂

**Comment by chrisblossom** [comment]
> It won't work with import as they're hoisted to the top level.

Do you mean `jest.doMock` does not work with `import`?

**Comment by SimenB** [comment]
Correct. See https://babeljs.io/en/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=FYUwzgLgdAJg9gWTgYwNYAoDkAjAhgJ0wEoBuAKDIEsBbABznwgAIBvAMzjgF8m385qTHAUzkyHOOlJA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=unambiguous&lineWrap=true&presets=env&prettier=true&targets=Node-6&version=7.4.5&externalPlugins=

If you use `jest.mock`, we ensure that the mock statement is hoisted above the transpiled `import`

**Comment by chrisblossom** [comment]
Yeah, but I've ran into countless issues because of the hoisting. Here is an example repo demonstrating some of them, and how I've fixed them:

[chrisblossom/issues/tree/eslint-plugin-jest-issue-283](https://github.com/chrisblossom/issues/tree/eslint-plugin-jest-issue-283)

Basically the only combination I've found to constantly work as expected is to use a combination of `jest.doMock` and dynamic requires (one day I'd like to write a babel plugin that does this automatically if possible). If for some reason I should open an issue on the jest repo let me know.

**Comment by SimenB** [comment]
This diff makes `example-mock.test.js` pass

```diff
diff --git i/src/example-mock.test.js w/src/example-mock.test.js
index 3c7fb08..76021dc 100644
--- i/src/example-mock.test.js
+++ w/src/example-mock.test.js
@@ -1,4 +1,3 @@
-import fs from 'fs';
 import path from 'path';
 import { readDummyFile } from './example';
 
@@ -6,7 +5,7 @@ const readDummyFileDynamic = () => require('./example').readDummyFile();
 
 function actuallyReadDummyFile() {
 	const pathname = path.resolve(__dirname, 'dummy-file.js');
-	const contents = fs.readFileSync(pathname, 'utf8');
+	const contents = jest.requireActual('fs').readFileSync(pathname, 'utf8');
 
 	return contents.trim();
 }
```

Does that cover your use case?

**Comment by chrisblossom** [comment]
Thanks for the response!

Yes, that will fix this minimal example, but that does not work if you are using an [npm package](https://github.com/chrisblossom/temp-sandbox) inside your test that uses `fs` (I've even [run into the case](https://github.com/chrisblossom/backtrack/blob/cb47bdb3bd727a94a8bc67a2e0f3312d2acde5be/src/stats-file/write-stats-file.test.ts#L24-L30) where I needed to restore `fs.writeFileSync` before calling `.toMatchSnapshot()` so jest could write a snapshot, is this a bug?).

I actually use `jest.requireActual` all the time (typically inside a `doMock`). I should have mentioned that.

I've [added an example](https://github.com/chrisblossom/issues/blob/eslint-plugin-jest-issue-283/src/read-file.test.js) where I don't see a way around using dynamic require.

Basically I've found instead of trying to troubleshoot why `jest.mock` isn't working, I just use a combination of `jest.doMock` and dynamic requires and it always just works as expected (especially with `resetModules: true`).

**Comment by SimenB** [comment]
That last example uses neither `jest.doMock` or `jest.mock`, though... It seems somewhat edge casey to me to want to both mock and not mock a dep within a single test.

What about using e.g. `jest.isolateModules`: https://jestjs.io/docs/en/jest-object#jestisolatemodulesfn? Hard to say if that solves your underlying issue without seeing some real code.

**Comment by chrisblossom** [comment]
> That last example uses neither `jest.doMock` or `jest.mock`, though...

Kind of got off topic from the original post (but they are directly related in my experience).

> It seems somewhat edge casey to me to want to both mock and not mock a dep within a single test.

I must live in edge cases, ha. ;) But really, I have been trying to show minimal examples, not real code. I'm sure there are multiple ways to solve this, but this pattern is the only way I've found to get everything "to just work" without having to think about the implementation details all the time. Which is why I'd like a `prefer-do-mock` rule.

> What about using e.g. `jest.isolateModules`?

I did not know about `jest.isolateModules`. I tried swapping out the dynamic require for it and it did not solve the issue.

**Comment by SimenB** [comment]
I'm not trying to be difficult, I'm just trying to understand if this is a problem we should tackle in Jest (via code and/or docs/examples) or if it's truly something we should provide a lint rule for. 🙂

Is there some open source test you can point to that you had to use `doMock` since `mock` didn't work? I'd like to understand the problem properly

**Comment by chrisblossom** [comment]
> I'm not trying to be difficult

Sorry if I came across in a way that made you think that. That never crossed my mind...quite the opposite actually, you are very helpful.

> Is there some open source test you can point to that you had to use doMock since mock didn't work? I'd like to understand the problem properly

The most recent example is an [attempt to refactor cosmiconfig](https://github.com/davidtheclark/cosmiconfig/pull/205#discussion_r303219220):

[Here is the original test code](https://github.com/chrisblossom/cosmiconfig/blob/modernize-code-with-dynamic-req/test/caches.test.ts). When using [`util.promisfy`](https://github.com/chrisblossom/cosmiconfig/blob/modernize-code-with-dynamic-req/src/readFile.ts#L2-L4) I can't find a way around using `doMock` in combination with dynamic requires ([it currently uses `jest.mock`](https://github.com/davidtheclark/cosmiconfig/blob/47d8951caf1ed5730e948d0aa0c65309eb53208f/test/caches.test.ts) and [a custom `Promise` wrapper](https://github.com/davidtheclark/cosmiconfig/blob/47d8951caf1ed5730e948d0aa0c65309eb53208f/src/readFile.ts#L14-L25)).


Here is another example I dug up: [chrisblossom/backtrack/src/clean/clean.test.ts](https://github.com/chrisblossom/backtrack/blob/jest.mock/src/clean/clean.test.ts)

**Comment by G-Rath** [comment]
From the comments above, it sounds like there are too many edge-cases for us to be able to implement a rule for this - you should be able to enforce this easily yourself with [`no-restricted-syntax`](https://eslint.org/docs/rules/no-restricted-syntax):

```
CallExpression[arguments.length > 0] > MemberExpression[object.name = 'jest'] > Identifier[name = 'mock']
```

**Comment by G-Rath** [comment]
Am closing per my previous comment
